crates/api/src/community/follow.rs

```diff
@@ -23,40 +24,52 @@ pub async fn follow_community(
   context: Data<LemmyContext>,
   local_user_view: LocalUserView,
 ) -> LemmyResult<Json<CommunityResponse>> {
+  check_user_valid(&local_user_view.person)?;
   let community = Community::read(&mut context.pool(), data.community_id).await?;
-  let mut community_follower_form = CommunityFollowerForm {
-    community_id: community.id,
-    person_id: local_user_view.person.id,
-    pending: false,
-  };
+  let form = CommunityFollowerForm::new(community.id, local_user_view.person.id);
 
   if data.follow {
+    // Only run these checks for local community, in case of remote community the local
+    // state may be outdated. Can't use check_community_user_action() here as it only allows
+    // actions from existing followers for private community (so following would be impossible).
     if community.local {
-      check_community_user_action(&local_user_view.person, community.id, &mut context.pool())
+      check_community_deleted_removed(&community)?;
+      CommunityPersonBanView::check(&mut context.pool(), local_user_view.person.id, community.id)
         .await?;
+    }
 
-      CommunityFollower::follow(&mut context.pool(), &community_follower_form)
-        .await
-        .with_lemmy_type(LemmyErrorType::CommunityFollowerAlreadyExists)?;
+    let state = if community.local {
+      // Local follow is accepted immediately
+      Some(CommunityFollowerState::Accepted)
+    } else if community.visibility == CommunityVisibility::Private {
+      // Private communities require manual approval
+      Some(CommunityFollowerState::ApprovalRequired)
```

Comment:
Is it possible for a community to be both local and private? Or remote and private?